File: backup/src/jclang/IndexLocation.java


package jclang;

import com.sun.jna.ptr.IntByReference;
import com.sun.jna.ptr.PointerByReference;
import org.jetbrains.annotations.NotNull;
import jclang.structs.CXIdxLoc;

import java.io.File;

public class IndexLocation {
    private final CXIdxLoc.ByValue location;

    private boolean isComputed;
    private File file;
    private int line;
    private int column;
    private int offset;

    public IndexLocation(@NotNull CXIdxLoc.ByValue location) {
        this.location = location;
    }

    private void retrieveLocation() {
        if (isComputed) return;
        isComputed = true;

        PointerByReference file = new PointerByReference();
        IntByReference line = new IntByReference();
        IntByReference column = new IntByReference();
        IntByReference offset = new IntByReference();
        LibClang.I.indexLoc_getFileLocation(location, null, file, line, column, offset);

        CXFile cxFile = new CXFile();
        cxFile.setPointer(file.getValue());
        this.file = cxFile.toFile();
        this.line = line.getValue();
        this.column = column.getValue();
        this.offset = offset.getValue();
    }

    // TODO: probably is null for unsaved files, investigate
    @NotNull
    public File getFile() {
        retrieveLocation();
        return file;
    }

    public int getLine() {
        retrieveLocation();
        return line;
    }

    public int getColumn() {
        retrieveLocation();
        return column;
    }

    public int getOffset() {
        retrieveLocation();
        return offset;
    }

    @NotNull
    public String toString() {
        return getFile() + "[" + getOffset() + "]:" + getLine() + ":" + getColumn();
    }
}

File: libclang/src/main/kotlin/klang/jvm/IndexLocation.kt
package klang.jvm

import com.sun.jna.ptr.IntByReference
import com.sun.jna.ptr.PointerByReference
import klang.jvm.binding.CXFile
import klang.jvm.binding.CXIdxLoc
import java.io.File

class IndexLocation(private val location: CXIdxLoc.ByValue) {
    private var isComputed = false
    private var file: File? = null
    private var line = 0
    private var column = 0
    private var offset = 0
    private fun retrieveLocation() {
        if (isComputed) return
        isComputed = true
        val file = PointerByReference()
        val line = IntByReference()
        val column = IntByReference()
        val offset = IntByReference()
        Clang.indexLoc_getFileLocation(location, null, file, line, column, offset)
        val cxFile = CXFile()
        cxFile.pointer = file.value
        this.file = cxFile.toFile()
        this.line = line.value
        this.column = column.value
        this.offset = offset.value
    }

    // TODO: probably is null for unsaved files, investigate
    fun getFile(): File {
        retrieveLocation()
        return file!!
    }

    fun getLine(): Int {
        retrieveLocation()
        return line
    }

    fun getColumn(): Int {
        retrieveLocation()
        return column
    }

    fun getOffset(): Int {
        retrieveLocation()
        return offset
    }

    override fun toString(): String {
        return getFile().toString() + "[" + getOffset() + "]:" + getLine() + ":" + getColumn()
    }
}